flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -1277,10 +1291,14 @@ public final ExecutorService getAsyncOperationsThreadPool() {
     private void notifyCheckpointComplete(long checkpointId) throws Exception {
         subtaskCheckpointCoordinator.notifyCheckpointComplete(
                 checkpointId, operatorChain, this::isRunning);
-        if (isRunning && isSynchronousSavepointId(checkpointId)) {
-            finishTask();
-            // Reset to "notify" the internal synchronous savepoint mailbox loop.
-            resetSynchronousSavepointId(checkpointId, true);
+        if (isRunning) {
+            if (isCurrentSavepointWithoutDrain(checkpointId)) {
+                finishTask();
+                // Reset to "notify" the internal synchronous savepoint mailbox loop.
+                syncSavepointWithoutDrain = null;
+            } else if (isCurrentSavepointWithDrain(checkpointId)) {
+                savepointCompletedFuture.complete(null);
+            }
```

Comment:
Do we have a ticket for unifying this behaviour as per offline discussion?

Reply:
Yes, we do: https://issues.apache.org/jira/browse/FLINK-23532